src/happy.rs

```diff
@@ -0,0 +1,84 @@
+//! aaa
+use std::io;
+use std::iter::FusedIterator;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::channel;
+use std::sync::Arc;
+use std::thread;
+use std::time::{Duration, Instant};
+
+fn intertwine<T, A, B>(mut ita: A, mut itb: B) -> Vec<T>
+where
+    A: FusedIterator<Item = T>,
+    B: FusedIterator<Item = T>,
+{
+    let mut res = vec![];
+
+    loop {
+        match (ita.next(), itb.next()) {
+            (Some(a), Some(b)) => {
+                res.push(a);
+                res.push(b);
+            }
+            (Some(a), None) => {
+                res.push(a);
+            }
+            (None, Some(b)) => {
+                res.push(b);
+            }
+            (None, None) => break,
+        }
+    }
+
+    res
+}
+
+/// aaa
+pub fn connect<A>(addrs: A) -> io::Result<TcpStream>
+where
+    A: ToSocketAddrs,
+{
+    let addrs: Vec<_> = addrs.to_socket_addrs()?.collect();
```

Comment:
`intertwine` seem quite general for this particular case and by rewriting it so that it handles a single `ToSocketAddrs` input directly seems like it could avoid the first `collect` call? IIRC the RFC also suggest to prefer IPv6 whereas this would spawn the connection to the first IPv4 address first?

Reply:
> intertwine seem quite general for this particular case and by rewriting it so that it handles a single ToSocketAddrs input directly seems like it could avoid the first collect call?

I went with a collect here so I could create two iterators on the same list of addresses. `to_socket_addrs` returns an iterator, I put then in a vec to avoid having to call `to_socket_addrs` twice which would perform DNS resolution twice. 

> IIRC the RFC also suggest to prefer IPv6 whereas this would spawn the connection to the first IPv4 address first?

The RFC says that you can sort addresses IPv4 first if you like. I figured that in the typical use case of the library, which is server apps and desktop apps, IPv4 would probably work better. IPv6 is popular in mobile networks but I don't see this library being used in mobile apps.

Reply:
> I put then in a vec to avoid having to call to_socket_addrs twice which would perform DNS resolution twice.

You are right, there is no avoiding the first collection. Maybe prefer `copied` over `cloned` to make it explicit that this is cheap? One could even write this as an iterator adaptor I guess to avoid the second collection, but I am not sure if it makes this clearer, e.g.
```
fn intertwine<T, A, B>(mut ita: A, mut itb: B) -> impl Iterator<Item = T>
where
    A: FusedIterator<Item = T>,
    B: FusedIterator<Item = T>,
{
    let mut stashed = None;

    std::iter::from_fn(move || {
        if stashed.is_some() {
            return stashed.take();
        }

        match (ita.next(), itb.next()) {
            (Some(a), Some(b)) => {
                stashed = Some(b);
                Some(a)
            }
            (Some(a), None) => Some(a),
            (None, Some(b)) => Some(b),
            (None, None) => None,
        }
    })
}
```

> IPv6 is popular in mobile networks but I don't see this library being used in mobile apps.

Personally, I would like it if IPv6 was preferred to generally push everybody to use IPv6 wherever it becomes available.

Reply:
Ok, it's not a big deal to me. My ISP does not support IPv6 right now, so I tried it out configured with IPv6 first and it doesn't seem to delay me much. The IPv6 connections fails quickly and IPv4 takes over.

Reply:
Thanks! (My ISP supports IPv6 as well, so things make not seem as bleak as they used to w.r.t. IPv6 deployment.)

(I think the iterator adaptor version of `intertwine` from above would still be nice)